### PR TITLE
Do not record unstaged deleted file upon recursive merge if file was …

### PR DESCRIPTION
**Pull request by mfilippov**
…moved outside of working tree with enabled sparse-checkout.

It's a very unexpected behaviour when a user sees a deleted file after a merge with enabled sparse-checkout. Moreover, when the user resolves merge conflicts and commits the changes with the command "git commit -am xxx", a repository can be broken because all the moved files will be deleted. Finally, it's really hard to find a user who deleted these files because "git log file" doesn't show any merge commits by default. I'm not sure that my fix is correct but I checked all tests and I didn't find a better way to prevent files deleting.

### CODE REVIEW

**Comment by shalupov** [comment]
Any news on this? I'm interested in this PR too

**Comment by hypersw** [comment]
This can be more or less ignored when working from clients such as GitEx or command line, but IDE integration in IDEA totally kills the repository because it stages all of the fake deletes automatically. So I'm very interested in this working correctly.

Any news on the issue?

**Comment by k15tfu** [comment]
@gitster could you look into this issue please?

**Comment by mfilippov** [comment]
PR contains test to reproduce the problem.

**Comment by mtrout** [comment]
I’m fairly new to all this obviously.  I’m really not sure what you mean.  There is quit the story behind why I’m here but I need to learn quickly how to recover control over my devices and network?  Can anyone help me?

File: merge-recursive.c [hunk 1]
```diff
@@ -1724,9 +1724,12 @@ static int merge_content(struct merge_options *o,
...
 		if (!path_renamed_outside_HEAD) {
-			add_cacheinfo(o, mfi.mode, &mfi.oid, path,
-				      0, (!o->call_depth), 0);
-			return mfi.clean;
+			struct stat st;
+			if (lstat(path, &st) == 0) {
+				add_cacheinfo(o, mfi.mode, &mfi.oid, path,
+					      0, (!o->call_depth), 0);
+				return mfi.clean;
+			}
```
Comment (dscho) [inline merge-recursive.c]: Unfortunately, the code has been changed quite a lot in the meantime. Would you mind rebasing it (and verifying that the problem you are fixing here has not been fixed in a different way already)?

File: t/t6042-merge-rename-corner-cases.sh [hunk 2]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
 	test ! -f c
 '
 
+test_expect_success 'move file/sparse-checkout/merge should not delete moved file' '
+	git rm -rf . &&
+	git clean -fdqx &&
+	rm -rf .git &&
+	git init &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: This is unnecessarily expensive. If you want a fresh worktree, just use `git init sparse-test`

File: t/t6042-merge-rename-corner-cases.sh [hunk 3]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	echo output >.gitignore &&
+	echo .gitignore >>.gitignore &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: Better: `echo output >>.git/info/exclude`

File: t/t6042-merge-rename-corner-cases.sh [hunk 4]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	echo b1 >b1 &&
+	git add b1 &&
+	git commit -m b1 &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: Better: `test_commit b1`. If you follow my advice to initialize a separate worktree for this test case, do `test_commit -C sparse-test b1`.

File: t/t6042-merge-rename-corner-cases.sh [hunk 5]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	mkdir excluded &&
+	echo problem >excluded/to-be-moved.txt &&
+	git add excluded/to-be-moved.txt &&
+	git commit -m to-be-moved &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: I think this would work, too, and be much shorter: `test_commit excluded/to-be-moved`

File: t/t6042-merge-rename-corner-cases.sh [hunk 6]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+	git tag split_point &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: If you use `test_commit`, there is no need to tag explicitly, as it has been tagged implicitly already.

It would make the code more to the point, easier to review, and easier to keep bug-free.

File: t/t6042-merge-rename-corner-cases.sh [hunk 7]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	echo b2 >b2 &&
+	git add b2 &&
+	git commit -m b2 &&
+	git tag b2 &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: `test_commit b2`

File: t/t6042-merge-rename-corner-cases.sh [hunk 8]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+	git reset --hard split_point &&
+
+	git mv excluded/to-be-moved.txt excluded/moved.txt &&
+	git commit -m move &&
+	git tag b1 &&
+
+	git config core.sparsecheckout true &&
+	echo "/*" >.git/info/sparse-checkout &&
+	echo "!excluded/" >>.git/info/sparse-checkout &&
+	git read-tree -mu HEAD &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: Would `git checkout &&` also do the job? You only want the `sparse-checkout` patterns to be applied, right?

File: t/t6042-merge-rename-corner-cases.sh [hunk 9]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+	git commit -m b2 &&
+	git tag b2 &&
+
+	git reset --hard split_point &&
+
+	git mv excluded/to-be-moved.txt excluded/moved.txt &&
+	git commit -m move &&
+	git tag b1 &&
+
+	git config core.sparsecheckout true &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: `test_config core.sparsecheckout true`.

File: t/t6042-merge-rename-corner-cases.sh [hunk 10]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	echo b2 >b2 &&
+	git add b2 &&
+	git commit -m b2 &&
+	git tag b2 &&
+
+	git reset --hard split_point &&
+
+	git mv excluded/to-be-moved.txt excluded/moved.txt &&
+	git commit -m move &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: In this instance, you indeed have to use `git commit`, as `test_commit` would not do what you want here.

File: t/t6042-merge-rename-corner-cases.sh [hunk 11]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+	git tag b1 &&
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: I don't see `b1` used anywhere... So why tag it?

And *if* this is tagged, maybe use a more informative name?

File: t/t6042-merge-rename-corner-cases.sh [hunk 12]
```diff
@@ -575,4 +575,45 @@ test_expect_success 'rename/rename/add-dest merge still knows about conflicting
...
+
+	git config core.sparsecheckout true &&
+	echo "/*" >.git/info/sparse-checkout &&
+	echo "!excluded/" >>.git/info/sparse-checkout &&
+	git read-tree -mu HEAD &&
+
+	git merge -m merge b2 &&
+
+	git status >output &&
+	test_i18ngrep "nothing to commit" output
```
Comment (dscho) [inline t/t6042-merge-rename-corner-cases.sh]: That looks good.

I do not actually believe that you need a separate worktree, but you can use the exact same worktree that has been used by the previous test cases already. If you are unsure whether it is clean (or if you are sure that it is not), you can just start the test case by `git reset --hard &&`.

**Comment by dscho** [comment]
@mfilippov gentle ping?

**Comment by mfilippov** [comment]
@dscho I trying to debug problem again. The test is actual but source significantly changed and my fix isn't working.

**Comment by mfilippov** [comment]
@dscho Maybe you knew, how can I run a single test?

**Comment by mfilippov** [comment]
@dscho I found how to do it. I move my test based on master and issue looks fixed.

**Comment by dscho** [comment]
> how can I run a single test?

By switching to the `t` directory and running the shell script directly.

> I found how to do it. I move my test based on master and issue looks fixed.

Cool!
